crypto-sodium/lib/Crypto/Sodium/Salt.hs

```diff
@@ -0,0 +1,81 @@
+-- SPDX-FileCopyrightText: 2021 Serokell
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | This module gives different ways of obtaining salts.
+--
+-- A “salt” is additional input provided to certain functions.
+-- Unlike a nonce, salt is used for “namespacing”, so there is not
+-- requirement that it is used only once, instead you use different
+-- salts to deisgnate different applications of an algorithm.
+--
+-- This is a subtle and pure semantical difference, so, for convenience,
+-- we also re-export some functions from "Crypto.Sodium.Nonce".
+module Crypto.Sodium.Salt
+  (
+  -- * Literals
+    utf8
+
+  -- * Random salt generation
+  , Crypto.Sodium.Nonce.generate
+  ) where
+
+import Data.ByteArray.Sized (SizedByteArray, unsafeSizedByteArray)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Unsafe as BS
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import GHC.Exts (Addr#)
+import Language.Haskell.TH (Exp, Q, Type)
+import qualified Language.Haskell.TH.Lib as TH
+import Language.Haskell.TH.Quote (QuasiQuoter (..))
+import Language.Haskell.TH.Syntax.Compat (SpliceQ)
+import qualified Language.Haskell.TH.Syntax.Compat as TH
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+import qualified Crypto.Sodium.Nonce
+
+
+-- | Quasi-quoter to construct a /sized/ 'ByteString' literal.
+--
+-- @
+-- {-# LANGUAGE QuasiQuotes #-}
+--
+-- {- ... -}
+--
+-- let salt = [utf8|hello world|] :: 'SizedByteArray' 11 ByteString
+-- @
+--
+-- This uses Template Haskell to compute the length of the UTF-8 encoding
+-- of the string you provide. Note that the string will be taken as is,
+-- with all whitespace preserved, for example @[utf8| x |]@ will have length 3.
+utf8 :: QuasiQuoter
+utf8 = QuasiQuoter { quoteExp, quotePat, quoteType, quoteDec }
+  where
+    quoteExp :: String -> Q Exp
+    quoteExp str = [e|unsafeSizedByteArray $(TH.unTypeSplice bsSplice) :: $tSized|]
+      where
+        bs :: ByteString
+        bs = T.encodeUtf8 $ T.pack str
+
+        len :: Int
+        len = BS.length bs
+
+        cstrSplice :: SpliceQ Addr#
+        cstrSplice = TH.unsafeSpliceCoerce . TH.litE . TH.stringPrimL . BS.unpack $ bs
+
+        bsSplice :: SpliceQ ByteString
+        bsSplice = [e||unsafeDupablePerformIO $ BS.unsafePackAddressLen len $$cstrSplice||]
```

Comment:
... I'll be frank. I think all the mess required for this single typed quote to work isn't worth it. Consider:

1. You had to introduce an additional dependency
2. You lost compatibility with GHC<8.10
3. The code needs coercions/conversions to/from typed splices now, making it less readable.

All that for the sake of a single line, basically, which, by the way, doesn't really improve type-safety all that much, because it's immediately used inside an untyped quote anyway.

I understand the intention behind preferring typed quotes in general, but here it seems the only reason to use the typed quote here is only for the sake of using a typed quote and nothing else.

Reply:
Yes, I thought about this...

I mean, I agree with your points, but the thing is, I don’t really care about compatibility with GHC 8.8, and I don’t think this extra dependency matters a lot. It is true that typed quotes are used here only for the sake of using typed quotes, however, I guess, this is kinda indeed the point?

There is quite a bit of unsafe stuff and C addresses going on, so having this part type-checked actually gives me some piece of mind and actually, I believe, this made it easier to write the code, so after weighing pros and cons I decided that this does improve the maintainability (if changes need to be made to this part in the future).

Also, I would not call this a mess? It is just two casts (free at runtime), one lightweight dependency (that will end up in the dependency closure sooner or later anyway), and the loss of GHC 8.8 (which I never cared about in the first place). So, honestly, I’m fine with this. If this causes any problems in the future, reverting this back to untyped TH is a matter of seconds.